source/Attribute.cpp

```diff
@@ -0,0 +1,425 @@
+		categoryName += "ing";
+	else if(category == SHIELD_GENERATION && effectType != SHIELDS)
+		categoryName = "shield";
+	else if(category == HULL_REPAIR && effectType != HULL)
+		categoryName = "hull";
+	else if(category == HULL_REPAIR && (effect == HULL || effect == HULL + 2 * ATTRIBUTE_EFFECT_COUNT))
+		categoryName = "hull repair rate";
+	else if(category == HULL_REPAIR && effectType == HULL)
+		categoryName = "hull repair";
+	else if(category == AFTERBURNING)
+		categoryName = "afterburner";
+	else if(category == RESISTANCE && secondaryName != "")
+		categoryName += " " + secondaryName;
+	else if(category == DAMAGE && effectType == PIERCING)
+		categoryName = "";
+	else if(category == DAMAGE && effectType == FORCE)
+		categoryName = "hit";
+	else if(category == ACTIVE_COOL && effectType != ACTIVE_COOLING)
+		categoryName = "cooling";
+	else if(category == DAMAGE && effectType == SCRAMBLE)
+		effectName = "scrambling";
+	if(effectType == LEAK && (category == THRUSTING || category == REVERSE_THRUSTING || category == TURNING
+			|| category == AFTERBURNING))
+		effectName = "leakage";
+	if((category >= PROTECTION && category <= DAMAGE) && effectType == SHIELDS)
+		effectName = "shield";
+	if(category == HULL_REPAIR && effectType == DELAY)
+		return "repair " + effectName;
+	if(category == HULL_REPAIR && effectType == DEPLETED_DELAY)
+		return "disabled repair delay";
+	if(category == SHIELD_GENERATION && effectType == DELAY)
+		return "shield " + effectName;
+	if(category == SHIELD_GENERATION && effectType == DEPLETED_DELAY)
+		return "depleted shield delay";
+	if(category == SHIELD_GENERATION && effectType == HULL)
+		effectName = "repair";
+
+	string composite;
+	// shortcuts: "energy generation energy" is "energy generation", etc.
+	if(static_cast<int>(category) == effectType && category <= CLOAKING)
+		composite = categoryName;
+	else if(category == PROTECTION || category == RESISTANCE || (category == DAMAGE && effect != FORCE))
+		composite = effectName + ((effectName == "" || categoryName == "") ? "" : " ") + categoryName;
+	else
+		composite = categoryName + ((effectName == "" || categoryName == "") ? "" : " ") + effectName;
+
+	if(IsRelative())
+		composite = "relative " + composite;
+	if(IsMultiplier())
+		composite += " multiplier";
+	return composite;
+}
+
+
+
+namespace {
+	// Cached mappings between the two formats. The indexing of newToOld is offset by one compared to the enum values.
+	string newToOld[ATTRIBUTE_CATEGORY_COUNT + 1][ATTRIBUTE_EFFECT_COUNT * 4 + 1][ATTRIBUTE_EFFECT_COUNT * 4 + 1];
+	unordered_map<string, Attribute*> oldToNew;
+}
+
+
+
+// Maps the attribute to the legacy single string format.
+string Attribute::GetLegacyName() const
+{
+	if(!cached)
+		PrepareCache();
+	return newToOld[category + 1][effect + 1][secondary + 1];
+}
+
+
+
+// Gets the attribute for the specified token, if any.
+Attribute *Attribute::Parse(const string &token)
+{
+	if(!cached)
+		PrepareCache();
+	auto it = oldToNew.find(token);
```

Comment:
```suggestion
	auto &it = oldToNew.find(token);
```
I think we can do this just fine?

Reply:
```
error: cannot bind non-const lvalue reference of type ‘std::__detail::_Node_iterator<std::pair<const std::__cxx11::basic_string<char>, Attribute*>, false, true>&’ to an rvalue of type ‘std::unordered_map<std::__cxx11::basic_string<char>, Attribute*>::iterator’ {aka ‘std::__detail::_Insert_base<std::__cxx11::basic_string<char>, std::pair<const std::__cxx11::basic_string<char>, Attribute*>, std::allocator<std::pair<const std::__cxx11::basic_string<char>, Attribute*> >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char> >, std::hash<std::__cxx11::basic_string<char> >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::iterator’}
```
Apparently not.